File: app/lib/data/models/monetization.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'monetization.freezed.dart';
part 'monetization.g.dart';

/// App credits per render quality, as the SERVER will charge them.
@freezed
abstract class RenderCosts with _$RenderCosts {
  const factory RenderCosts({
    @Default(1) int standard,
    @Default(4) int hd,
    @Default(4) int enhance,
  }) = _RenderCosts;

  factory RenderCosts.fromJson(Map<String, dynamic> json) =>
      _$RenderCostsFromJson(json);
}

/// One purchasable plan as the server describes it.
///
/// `priceUsd` is for reconciliation ONLY. The paywall must render the STORE's
/// localized price string (RETENTION spec §7.2) — the store is the authority on
/// what the user actually pays, and this number is not localized, not taxed and
/// not currency-converted.
@freezed
abstract class PlanPresentation with _$PlanPresentation {
  const factory PlanPresentation({
    required String tier,
    @Default('subscription') String kind,
    @JsonKey(name: 'price_usd') @Default(0.0) double priceUsd,
    @JsonKey(name: 'monthly_credits') @Default(0) int monthlyCredits,
    @JsonKey(name: 'hd_allowed') @Default(false) bool hdAllowed,
    @Default(false) bool priority,
    @JsonKey(name: 'play_product_id') String? playProductId,
    @JsonKey(name: 'app_product_id') String? appProductId,
  }) = _PlanPresentation;

  factory PlanPresentation.fromJson(Map<String, dynamic> json) =>
      _$PlanPresentationFromJson(json);
}

/// When WTM is allowed to interrupt the user with a monetization surface.
@freezed
abstract class PaywallPolicy with _$PaywallPolicy {
  const factory PaywallPolicy({
    @JsonKey(name: 'cooldown_hours') @Default(24) int cooldownHours,
    @JsonKey(name: 'post_purchase_cooldown_hours')
    @Default(72)
    int postPurchaseCooldownHours,
    @JsonKey(name: 'timing_variant') @Default('control') String timingVariant,

    /// False when a cooldown is in force. Consulted ONLY by surfaces WTM
    /// raises on its own — a paywall the user opened never checks this.
    @JsonKey(name: 'may_interrupt') @Default(true) bool mayInterrupt,
    @JsonKey(name: 'block_reason') String? blockReason,
    @JsonKey(name: 'retry_after_hours') int? retryAfterHours,
  }) = _PaywallPolicy;

  factory PaywallPolicy.fromJson(Map<String, dynamic> json) =>
      _$PaywallPolicyFromJson(json);
}

/// The server-authoritative monetization snapshot for this user.
///
/// The app REFLECTS this; it never enforces from it. Every cost, gate and
/// entitlement is applied server-side, so a client that ignored this response
/// entirely would face the same charges and the same refusals.
@freezed
abstract class MonetizationConfig with _$MonetizationConfig {
  const factory MonetizationConfig({
    @Default(1) int version,
    @JsonKey(name: 'render_costs')
    @Default(RenderCosts())
    RenderCosts renderCosts,
    @JsonKey(name: 'free_render_limit') @Default(0) int freeRenderLimit,
    @JsonKey(name: 'free_render_remaining') @Default(0) int freeRenderRemaining,
    @Default('free') String tier,
    @JsonKey(name: 'hd_allowed') @Default(false) bool hdAllowed,
    @Default(<PlanPresentation>[]) List<PlanPresentation> plans,
    @Default(PaywallPolicy()) PaywallPolicy paywall,
    @JsonKey(name: 'trial_enabled') @Default(false) bool trialEnabled,
    @JsonKey(name: 'trial_credit_cap') int? trialCreditCap,
    @JsonKey(name: 'rollover_enabled') @Default(false) bool rolloverEnabled,
    @Default(<String, String>{}) Map<String, String> experiments,
    @JsonKey(name: 'paywall_v2') @Default(false) bool paywallV2,
    @JsonKey(name: 'render_gate_v2') @Default(false) bool renderGateV2,
  }) = _MonetizationConfig;

  const MonetizationConfig._();

  factory MonetizationConfig.fromJson(Map<String, dynamic> json) =>
      _$MonetizationConfigFromJson(json);

  bool get isSubscriber => tier != 'free';

  /// True when a free user has spent their lifetime renders. Deliberately NOT
  /// a reason to lock the app: saved looks, planning and Style Memory all stay
  /// open (§9) — only a NEW paid render is gated.
  bool get freeRendersExhausted => !isSubscriber && freeRenderRemaining <= 0;
}

/// A monetization surface was shown, dismissed, acted on, or converted.
enum MonetizationSurface {
  paywall('paywall'),
  topupSheet('topup_sheet'),
  renderGate('render_gate'),
  upgradePrompt('upgrade_prompt');

  const MonetizationSurface(this.wire);

  final String wire;
}

enum MonetizationAction {
  viewed('viewed'),
  dismissed('dismissed'),
  ctaTapped('cta_tapped'),
  purchased('purchased');

  const MonetizationAction(this.wire);

  final String wire;
}

File: backend/app/models/monetization.py
"""Monetization config API models (spec §39).

The snapshot the app fetches so an experiment number can change without a
binary release. Everything here is DESCRIPTIVE — the server still enforces every
cost, gate and entitlement itself (§11). A client that ignored this response
entirely would get the same charges and the same refusals; it would just show
worse copy.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class RenderCosts(BaseModel):
    """App credits per render quality. Mirrors what the server WILL charge."""

    standard: int
    hd: int
    enhance: int


class PlanPresentation(BaseModel):
    """One purchasable plan as the server describes it.

    `price_usd` is the catalog price for reference and reconciliation ONLY. The
    app must display the STORE's localized price string (§7.2) — never this
    number — because the store is the authority on what the user actually pays.
    """

    tier: str
    kind: str
    price_usd: float
    monthly_credits: int
    hd_allowed: bool
    priority: bool
    play_product_id: str | None = None
    app_product_id: str | None = None


class PaywallPolicy(BaseModel):
    """When WTM is allowed to interrupt (§10)."""

    cooldown_hours: int
    post_purchase_cooldown_hours: int
    timing_variant: str
    #: False when a cooldown is currently in force. A surface the user opens
    #: themselves ignores this entirely.
    may_interrupt: bool = True
    block_reason: str | None = None
    retry_after_hours: int | None = None


class MonetizationConfig(BaseModel):
    """The whole policy snapshot for the calling user."""

    #: Bumped when the shape changes so an old client can tell.
    version: int = 1
    render_costs: RenderCosts
    #: Lifetime free standard renders for this user, as the server will enforce.
    free_render_limit: int
    free_render_remaining: int
    tier: str
    hd_allowed: bool
    plans: list[PlanPresentation] = Field(default_factory=list)
    paywall: PaywallPolicy
    trial_enabled: bool = False
    trial_credit_cap: int | None = None
    rollover_enabled: bool = False
    #: {experiment: variant}. Empty for everyone not in an experiment.
    experiments: dict[str, str] = Field(default_factory=dict)
    #: Which v2 surfaces are live for this user, so the app does not need a
    #: second round-trip to /v1/flags to compose a paywall.
    paywall_v2: bool = False
    render_gate_v2: bool = False


class MonetizationEventIn(BaseModel):
    """A monetization surface was shown, dismissed, acted on, or converted."""

    surface: str = Field(max_length=40)
    action: str = Field(pattern="^(viewed|dismissed|cta_tapped|purchased)$")
    #: True ONLY when WTM raised this surface on its own. A user-initiated
    #: paywall must send false, or it would start a cooldown against itself.
    interruptive: bool = False
    context: dict[str, str] | None = None
